lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -275,9 +277,134 @@ bool PatchEntryPointMutate::lowerCpsOps(Function *func) {
+  auto *vcr = builder.CreateExtractValue(vgprArg, 0);
+  auto *vcrTy = vcr->getType();
+
+  if (isCpsFunc) {
+    // FIXME: Switch to LLVM intrinsic
+    FunctionType *setInactiveChainTy = FunctionType::get(vcrTy, {vcrTy, vcrTy}, false);
+    auto setInactiveChain = Function::Create(setInactiveChainTy, GlobalValue::ExternalLinkage,
+                                             "llvm.amdgcn.setinactive.chain.arg", func->getParent());
+    vcr = builder.CreateCall(setInactiveChain, {vcr, func->getArg(numUserdata)});
+  }
+
+  // Find first lane having non-null vcr, and use as next jump target.
+  auto *vcrMask = builder.CreateICmpNE(vcr, builder.getInt32(0));
+  auto *pendingBallot = builder.CreateIntrinsic(Intrinsic::amdgcn_ballot, waveMaskTy, vcrMask);
+  Value *firstActive = builder.CreateIntrinsic(Intrinsic::cttz, waveMaskTy, {pendingBallot, builder.getTrue()});
+  if (!waveMaskTy->isIntegerTy(32))
+    firstActive = builder.CreateTrunc(firstActive, builder.getInt32Ty());
+  auto *targetVcr = builder.CreateIntrinsic(builder.getInt32Ty(), Intrinsic::amdgcn_readlane, {vcr, firstActive});
+  // Calculate the lane mask that take this specific target.
+  auto *targetMask = builder.CreateICmpEQ(vcr, targetVcr);
+  auto *execMask = builder.CreateIntrinsic(Intrinsic::amdgcn_ballot, waveMaskTy, targetMask);
+
+  if (isCpsFunc) {
+    targetVcr = builder.CreateUnaryIntrinsic(Intrinsic::amdgcn_wwm, targetVcr);
+    execMask = builder.CreateUnaryIntrinsic(Intrinsic::amdgcn_wwm, execMask);
+  }
+
+  auto *isNullTarget = builder.CreateICmpEQ(targetVcr, builder.getInt32(0));
+  builder.CreateCondBr(isNullTarget, retBlock, chainBlock);
+
+  builder.SetInsertPoint(retBlock);
+  builder.CreateRetVoid();
+
+  builder.SetInsertPoint(chainBlock);
+  // Mask off metadata bits and setup jump target.
+  Value *addr32 = builder.CreateAnd(targetVcr, builder.getInt32(~0x3fu));
+  AddressExtender addressExtender(func);
+  Value *jumpTarget = addressExtender.extend(addr32, builder.getInt32(HighAddrPc), builder.getPtrTy(), builder);
+
+  Value *chainArgs[] = {jumpTarget, execMask, userDataVec, vgprArg, builder.getInt32(0)};
+
+#if LLVM_MAIN_REVISION && LLVM_MAIN_REVISION < 465197
+  // Old version of the code
+  Type *chainArgTys[] = {builder.getPtrTy(), builder.getIntNTy(waveSize), userDataVec->getType(), vgprArg->getType(),
+                         builder.getInt32Ty()};
+  FunctionType *chainFuncTy = FunctionType::get(builder.getVoidTy(), chainArgTys, true);
+  auto chainFunc =
+      Function::Create(chainFuncTy, GlobalValue::ExternalLinkage, "llvm.amdgcn.cs.chain", func->getParent());
+  builder.CreateCall(chainFunc, chainArgs);
+#else
+  // New version of the code (also handles unknown version, which we treat as
+  // latest)
+  Type *chainTys[] = {builder.getPtrTy(), builder.getIntNTy(waveSize), userDataVec->getType(), vgprArg->getType()};
+  builder.CreateIntrinsic(Intrinsic::amdgcn_cs_chain, chainTys, chainArgs);
```

Comment:
Use the overload of `CreateIntrinsic` that doesn't require explicitly passing the overload set. It's the one where you explicitly pass `void` as return type, and then the rest should be inferred from `chainArgs`.

Reply:
The code was contributed by @jayfoad . Will that version mangle or not mangle the intrinsic name? I would leave the change later when the LLVM backend change was propagated to github. So I can replace it and update the test check as well.

Reply:
I tried, but it does not work as the other version of `CreateIntrinsic` has limitation in supporting vararg intrinsic. I would leave this as future work.

Reply:
> the other version of CreateIntrinsic has limitation in supporting vararg intrinsic

Sorry, I did try but I failed: https://reviews.llvm.org/D130776#3691034
It would be great if someone else could implement that.